feat: Handling pagination of repository vulnerabilities

JoseAngel1196 opened the pull request:
Fixes #7 

# Description

In this PR, we're addressing pagination for repository vulnerabilities. To put it simply, imagine If we have a substantial number of vulnerabilities for a repository (which may not be ideal), we need to ensure we handle them appropriately. It's also not ideal to skip these vulnerabilities, as we are currently doing today.

I have ensured that these new changes do not break the reporting functionality. I have verified this locally using a test account.

tarkatronic reviewed (changes requested):
Hmmm, there's a problem with this approach. This will drastically increase the number of calls made to the GraphQL API, which will both slow us down and potentially cause rate limiting. Previously, given `n = number of repositories`, we would perform `n / 100` API calls. Now we perform `n + 1` calls. Given a sufficiently large org, such as one containing 500 repositories, we just went from 5 calls to 501. And that's assuming we never actually paginate the vulnerabilities, which is the whole point here!

I think a better way we could accomplish this is through a combination of approaches. In other words, only perform these repo-specific queries when necessary.

Taking another look at this, I believe this could actually be somewhat easy using mostly just the pre-existing data structures. The current flow would work as-is, but then when `HasNextPage` is `true` for a repositories vulnerabilities, you would send a new query, utilizing the `EndCursor`, which would look something like this:

```go
type repoVulnerabilityQuery struct {
  Repository orgRepo `graphql:"repository(name: $repoName, owner: $orgName)"`
}
```

I think that might be the only new data structure necessary.. and then we would continue to have our `n / 100` queries, except in cases where vulnerabilities are paginated.

I feel like I may have rambled a bit in there, so I hope it all makes sense. Please feel free to ask if you have any questions!

JoseAngel1196 commented:
> Hmmm, there's a problem with this approach. This will drastically increase the number of calls made to the GraphQL API, which will both slow us down and potentially cause rate limiting. Previously, given `n = number of repositories`, we would perform `n / 100` API calls. Now we perform `n + 1` calls. Given a sufficiently large org, such as one containing 500 repositories, we just went from 5 calls to 501. And that's assuming we never actually paginate the vulnerabilities, which is the whole point here!
> 
> I think a better way we could accomplish this is through a combination of approaches. In other words, only perform these repo-specific queries when necessary.
> 
> Taking another look at this, I believe this could actually be somewhat easy using mostly just the pre-existing data structures. The current flow would work as-is, but then when `HasNextPage` is `true` for a repositories vulnerabilities, you would send a new query, utilizing the `EndCursor`, which would look something like this:
> 
> ```go
> type repoVulnerabilityQuery struct {
>   Repository orgRepo `graphql:"repository(name: $repoName, owner: $orgName)"`
> }
> ```
> 
> I think that might be the only new data structure necessary.. and then we would continue to have our `n / 100` queries, except in cases where vulnerabilities are paginated.
> 
> I feel like I may have rambled a bit in there, so I hope it all makes sense. Please feel free to ask if you have any questions!

Alright Joey, thanks for the feedback. I have addressed the comment, lmk what you think. Smart thinking about the rate limiting. Will keep this in mind for the future!

JoseAngel1196 commented on querying/github.go:
```diff
@@ -167,6 +173,24 @@ func (gh *GithubDataSource) processRepoFindings(projects *ProjectCollection, rep
+
+		nextCursor := repositoryQuery.Repository.VulnerabilityAlerts.PageInfo.EndCursor
+		log.Info().Any("alertCursor", queryVars["alertCursor"]).Msg("Querying for more vulnerabilities for a repository.")
+		return gh.processRepoFindings(projects, repositoryQuery.Repository, nextCursor)
```
I have verified this locally, will add some unit test for this in a follow up PR.

tarkatronic commented on querying/github.go:
```diff
@@ -137,12 +143,12 @@ func (gh *GithubDataSource) CollectFindings(projects *ProjectCollection, wg *syn
 }
 
-func (gh *GithubDataSource) processRepoFindings(projects *ProjectCollection, repo orgRepo, repoCursor *githubv4.String) error {
+func (gh *GithubDataSource) processRepoFindings(projects *ProjectCollection, repo orgRepo, endCursor githubv4.String) error {
```
🤔 Is it even necessary to pass the `endCursor` in now with this new structure? You should be able to access this directly with `repo.VulnerabilityAlerts.PageInfo.EndCursor`. That should work... right?

JoseAngel1196 replied:
yeah, you are RIGHT, totally my bad haha

JoseAngel1196 replied:
Fixed!

tarkatronic commented on querying/github.go:
```diff
@@ -167,6 +173,24 @@ func (gh *GithubDataSource) processRepoFindings(projects *ProjectCollection, rep
 	}
+
+	if repo.VulnerabilityAlerts.PageInfo.HasNextPage {
+		var repositoryQuery repositoryQuery
```
Careful! It looks like you are unintentionally shadowing this name here -- using the same name in an inner scope as in an outer scope. This can potentially cause issues! I would recommend using a different name for this variable, something like

```suggestion
		var repoQuery repositoryQuery
```
(Also updating line 184-191 to reflect this)
This ensures there will be no conflicts, confusion, or weirdness.

JoseAngel1196 replied:
Fixed

tarkatronic reviewed (commented):
This is great! I love the recursive approach. And this is so much easier than I had feared in my head!

I just wanted to quickly address these last couple of comments, but aside from that this looks all ready!

tarkatronic commented on querying/github.go:
```diff
@@ -167,6 +173,23 @@ func (gh *GithubDataSource) processRepoFindings(projects *ProjectCollection, rep
+			return err
+		}
+
+		log.Info().Any("alertCursor", queryVars["alertCursor"]).Msg("Querying for more vulnerabilities for a repository.")
```
```suggestion
		log.Info().Str("repoName", repo.Name).Any("alertCursor", queryVars["alertCursor"]).Msg("Querying for more vulnerabilities for a repository.")
```
Let's log the repo name as well, so we know what we're iterating on 😄

JoseAngel1196 replied:
Done!

tarkatronic reviewed (approved):
Amazing! Wonderful! Love it! Thanks so much for contributing this @JoseAngel1196
